engineering_tools/carma

```diff
@@ -339,6 +339,174 @@ carma__stop() {
+            docker stack deploy --compose-file - carma-pc1
+
+            docker run --rm -v carma-config-data:/opt/carma/vehicle/config --entrypoint sh busybox:latest -c \
+            'cat /opt/carma/vehicle/config/docker-compose-pc2.yml' | \
+            docker stack deploy --compose-file - carma-pc2
+            
+        
+        elif [ ! -z "$2" ]; then
+            echo "Unrecognized argument \"swarm deploy $2\""
+        fi
+
+
+    # Updates the carma-config-data volume based on a input tag string given by the user
+    elif [ "$1" = "update-config" ]; then
+        echo "Updating carma-config-data volume..."
+
+        if [ "$2" = "-d" ]; then
+            local USERNAME=usdotfhwastoldev
+            local TAG=$3
+        elif [ "$2" = "-c" ]; then
+            local USERNAME=usdotfhwastolcandidate
+            local TAG=$3
+        else
+            local USERNAME=usdotfhwastol
+            local TAG=$2
+        fi
+
+        if [[ -z $2 ]]; then
+            echo "Please specify a tag string to update carma-config-data volume."
+            echo "Done."
+            exit 1
+        fi
+
+        local IMAGE_NAME="$USERNAME/carma-config:$TAG"
+
+        if [ "$(docker images | grep $USERNAME/carma-config | grep $TAG )" ]; then
+            echo "Found image, applying to carma-config-volume"
+            docker run --rm --name carma-config-data -v carma-config-data:/opt/carma/vehicle/config -d $USERNAME/carma-config:$TAG
+        else
+            echo "Could not find $USERNAME/carma-config:$TAG in local images"
+        fi
+
+
+    elif [ "$1" = "edit-config" ]; then
+        echo "Opening carma-config-data volume shell with read/write privileges..."
+
+        if [ "$(docker volume ls | grep carma-config-data | sed 's/local     //g')" ]; then
+            echo "Found carma-config-data volume"
+        else
+            echo "Missing carma-config-data volume, add with \"carma swarm update-config <image tag>\""
+            exit 1
+        fi
+
+        local carma_base=$(__get_image_from_config carma-base:)
+        if [[ -z $carma_base ]]; then
+            __pull_newest_carma_base
+            carma_base=$(__get_most_recent_carma_base)
+        fi
+
+        docker run -it --rm -v carma-config-data:/opt/carma/vehicle/config $carma_base bash
+
+
+    # Attaches each carma service's STDOUT logs to the current terminal
+    elif [ "$1" = "attach" ]; then
+        echo "Attaching STDOUT logs to terminal..."
+
+        SWARM_SERVICES=$(docker service ls --format "{{.Name}}")
+
+        # this seems to loop forever and cannot be escaped with ctrl+C
+        # performance will likely be poor due to combined log streams
+        for TASK in $SWARM_SERVICES; do
+            docker service logs --follow --details --tail=10 "$TASK" &
+        done
+
+        sleep infinity
+
+
+    # Takes n image tags that are saved locally and pushes them to the registry
+    elif [ "$1" = "set-image" ]; then
```

Comment:
If `carma swarm set-image image:tag` pushes the specified image to the local registry, would the user be expected to then ssh into the worker PC and pull down the image that have been pushed from the manager PC? Just trying to check my understanding.

Reply:
Yes- I mentioned this in a previous registry comment to Kyle above and I @'d you for your thoughts